prow/kube/config.go

```diff
@@ -92,16 +83,96 @@ func LoadClusterConfigs(kubeconfig, projectedTokenFile string) (map[string]rest.
 	} else {
 		localCfg.UserAgent = version.UserAgent()
 	}
-	if localCfg != nil && projectedTokenFile != "" {
+	if localCfg != nil && kc.ProjectedTokenFile != "" {
 		localCfg.BearerToken = ""
-		localCfg.BearerTokenFile = projectedTokenFile
-		logrus.WithField("tokenfile", projectedTokenFile).Info("Using projected token file")
+		localCfg.BearerTokenFile = kc.ProjectedTokenFile
+		logrus.WithField("tokenfile", kc.ProjectedTokenFile).Info("Using projected token file")
 	}
 
-	kubeCfgs, currentContext, err := kubeConfigs(kubeconfig)
-	if err != nil {
-		return nil, fmt.Errorf("kubecfg: %v", err)
+	var candidates []string
+	if kc.File != "" {
+		candidates = append(candidates, kc.File)
+	}
+	if kc.Dir != "" {
+		files, err := ioutil.ReadDir(kc.Dir)
+		if err != nil {
+			return nil, fmt.Errorf("kubecfg dir: %v", err)
+		}
+		var re *regexp.Regexp
+		for _, file := range files {
+			filename := file.Name()
+			if re == nil {
+				re, err = regexp.Compile(kc.RegExRaw)
+				if err != nil {
+					return nil, fmt.Errorf("failed to compile regex from %q: %w", kc.RegExRaw, err)
+				}
+			}
+			if !re.MatchString(filename) {
+				continue
+			}
+			candidates = append(candidates, filepath.Join(kc.Dir, file.Name()))
+		}
 	}
 
-	return mergeConfigs(localCfg, kubeCfgs, currentContext)
+	allKubeCfgs := map[string]rest.Config{}
+	var currentContext string
+	if len(candidates) == 0 {
+		// loading from the defaults, e.g., ${KUBECONFIG}
+		if allKubeCfgs, currentContext, err = kubeConfigs(clientcmd.NewDefaultClientConfigLoadingRules()); err != nil {
+			logrus.WithError(err).Warn("Cannot load kubecfg")
+		}
+	} else {
+		for _, candidate := range candidates {
+			logrus.Infof("Loading kubeconfig from: %q", candidate)
+			kubeCfgs, tempCurrentContext, err := kubeConfigs(&clientcmd.ClientConfigLoadingRules{ExplicitPath: candidate})
+			if err != nil {
+				return nil, fmt.Errorf("fail to load kubecfg from %q: %v", candidate, err)
+			}
+			currentContext = tempCurrentContext
+			for c, k := range kubeCfgs {
+				if _, ok := allKubeCfgs[c]; ok {
+					return nil, fmt.Errorf("context %s occurred more than once in kubeconfig dir %q", c, kc.Dir)
+				}
+				allKubeCfgs[c] = k
+			}
+		}
+	}
+
+	return mergeConfigs(localCfg, allKubeCfgs, currentContext)
+}
+
+type KubeConfig struct {
```

Comment:
This stutters. https://blog.golang.org/package-names

Prefer `kube.Config`

Reply:
Why is this struct (and the fields) exported?

Reply:
Updated.